Tox

yannikschaelte opened the pull request:
Main: Add tox for easier testing. This allows to
* Run tests locally by just typing `tox` (or `tox -e pyroma` for that specific test)
* Have everything in well-defined environments. In particular, thus the flake8 plugins can be handled better and do not mess up with other projects were other plugins are used
* Configure the tests on GHA without any implementation overhead

Further:
* Move flake8 specifications to standard `.flake8` file (remove `run_flake8.sh`, can be replaced by `tox -e flake8`)
* Add notebook tests (nbmake) via `tox -e notebooks`
* Add further python conformity tests (pyroma)
* The coverage can now also be inspected locally better (`tox -e test; firefox htmlcov/index.html`)

yannikschaelte commented:
@jvanhoefer @martamatos I have here set up a rather easy to use testing system based on tox, and added the notebook tests. As pointed out in #67, several notebooks fail.

What do you think about tox (see above discussion)? If it is not desirable, most of the tests can also be written directly in GHA (for flake8 a little care needs to be taken). Open for discussion :)

yannikschaelte commented on .github/workflows/ci.yml:
```diff
@@ -1,6 +1,6 @@
 name: CI
 
-on: [push]
+on: [pull_request]
```
note: I think this makes it easier for external collaborators

jvanhoefer reviewed (approved):
Can we add some section in the docs how to run tests using tox? Besides that: Looks good!

yannikschaelte commented:
> Can we add some section in the docs how to run tests using tox? Besides that: Looks good!

Yes, sure! If we keep tox, I will update the docs now and add some info on how to run things also locally. Just didn't want to do too much in the first place.

yannikschaelte commented:
@jvanhoefer I have updated (and simplified / shortened) the docs now, and annotated the code. I suggest to disable the notebooks test until #90 is fully solved (1 notebook remains).

jvanhoefer commented:
regarding the testing: Can you restrict amici to version v0.11.10? The notebook works locally for me on that version and afaik there have been changes to piecewise functions in AMICI recently :)

yannikschaelte commented:
Fixed the AMICI version, now all notebooks pass. Added a TODO in setup.cfg to check this in the future. Good to go then from my side. :)
